# Exploratory Data Analysis (EDA)

Utilizando el dataset de Olist, una plataforma de comercio electrónico brasileña, se plantea la hipótesis de que es posible predecir qué clientes presentan un patrón de comportamiento asociado al abandono del servicio (churn).

Esta métrica es crucial para las empresas, ya que permite identificar de forma anticipada a los clientes en riesgo y aplicar estrategias de retención proactivas, lo cual puede traducirse en una mejora directa en los ingresos y en la fidelización a largo plazo.

El conjunto de datos abarca compras realizadas entre septiembre de 2016 y octubre de 2018, e incluye información detallada sobre los pedidos, pagos, entregas, localización del cliente y puntuaciones de reseñas.

Para efectos de este análisis, se define churn como aquellos clientes que no han realizado ninguna compra en los últimos 6 meses del periodo registrado. Esta ventana temporal es razonable dado el tipo de productos vendidos en la plataforma, que no exigen compras frecuentes ni suscripciones.

### Importado de librerias y dataset

In [1]:
from pathlib import Path

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
import folium

In [2]:
BASE_DIR = Path.cwd().parent
DATA_DIR = (BASE_DIR / "data").resolve()

df = pd.read_csv(DATA_DIR/"processed/processed_dataset.csv")

### Observación general

Lo primero es realizar una observación general del dataset

In [3]:
df.head()


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  total_payment payment_type  ...  \
0           2017-10-18 00:00:00          38.71      voucher  ...   
1           2018-08-13 00:00:00         141.46       boleto  ...   
2           2018-09-04 00:00:00         179.12  credit_card  ...   
3           2017-12-15 00:00:00          72.20  credit_card  ...   
4           2018-02-26 00:00:00          28.62  credit_card  ...   

   worst_review  mean_review_score  last_review  \
0             4                4.0            4   
1             4                4.0            4   
2             5                5.0            5   
3             5                5.0            5   
4             5                5.0            5   

                              review_comment_message  review_comment_title  \
0  Não testei o produto ainda, mas ele veio corre...                   NaN   
1                               Muito bom o produto.      Muito boa a loja   
2                                                NaN                   NaN   
3  O produto foi exatamente o que eu esperava e e...                   NaN   
4                                                NaN                   NaN   

  total_price item_count  total_freight_value            customer_city  \
0       29.99          1                 8.72                sao paulo   
1      118.70          1                22.76                barreiras   
2      159.90          1                19.22               vianopolis   
3       45.00          1                27.20  sao goncalo do amarante   
4       19.90          1                 8.72              santo andre   

   customer_state  
0              SP  
1              BA  
2              GO  
3              RN  
4              SP  

[5 rows x 22 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97916 entries, 0 to 97915
Data columns (total 22 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   order_id                       97916 non-null  object 
 1   customer_id                    97916 non-null  object 
 2   order_status                   97916 non-null  object 
 3   order_purchase_timestamp       97916 non-null  object 
 4   order_approved_at              97902 non-null  object 
 5   order_delivered_carrier_date   96925 non-null  object 
 6   order_delivered_customer_date  95829 non-null  object 
 7   order_estimated_delivery_date  97916 non-null  object 
 8   total_payment                  97916 non-null  float64
 9   payment_type                   97916 non-null  object 
 10  max_installments               97916 non-null  int64  
 11  total_reviews                  97916 non-null  int64  
 12  worst_review                   97916 non-null 

Las columnas de fechas, estan en formato object, hay que cambiarlas a datetime

In [5]:
date_cols = [
    'order_purchase_timestamp', 
    'order_approved_at', 
    'order_delivered_carrier_date', 
    'order_delivered_customer_date', 
    'order_estimated_delivery_date'
    ]

for col in date_cols:
    df[col] = pd.to_datetime(df[col])

Una vez hecho esto comprobare los datos nulos de mi dataset, por el preproceso de estos datos, se que hay datos nulos en las columnas de reviews, vamos a ver que mas columnas pueden tener datos nulos, que tenemos que tener en cuenta.

In [6]:
df.isnull().sum()

order_id                             0
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                   14
order_delivered_carrier_date       991
order_delivered_customer_date     2087
order_estimated_delivery_date        0
total_payment                        0
payment_type                         0
max_installments                     0
total_reviews                        0
worst_review                         0
mean_review_score                    0
last_review                          0
review_comment_message           57605
review_comment_title             86452
total_price                          0
item_count                           0
total_freight_value                  0
customer_city                        0
customer_state                       0
dtype: int64

## Declaracion de mi variable Churn o Target

Como he indicado anteriormente, se considerara como Churn(abandono del servicio), cuando un cliente no haya utilizado la plataforma en los ultimos 6 meses. De esta manera, evitamos calificar como Churn la mayoria de compras estacionales y podemos afinar mejor nuestra prediccion.

In [7]:
# Lo primero es obtener las ultimas compras de cada cliente, para ello hay que comprobar que sean unicos nuestros codigos de cliente
print('Hay', df['order_id'].duplicated().sum(), 'order_id duplicados')
print('Hay', df['customer_id'].duplicated().sum(), 'customer_id duplicados')

Hay 0 order_id duplicados
Hay 0 customer_id duplicados


In [ ]:
def churn_rate(df, days):
    last_purchase = df.groupby('customer_id')['order_purchase_timestamp'].max().reset_index()
    cutoff_date = df["order_purchase_timestamp"].max() - pd.Timedelta(days=days)
    last_purchase["churn"] = last_purchase["order_purchase_timestamp"] < cutoff_date
    churn_ratio = last_purchase["churn"].value_counts(normalize=True)
    print(f"\nVentana: {days} días")
    print(churn_ratio)

for days in [90, 180, 365]:
    churn_rate(df, days)


Ventana: 90 días
churn
True     0.815822
False    0.184178
Name: proportion, dtype: float64

Ventana: 180 días
churn
True     0.611024
False    0.388976
Name: proportion, dtype: float64

Ventana: 365 días
churn
False    0.764247
True     0.235753
Name: proportion, dtype: float64


Con esto tenemos la variable Churn establecida. 

Estas proporciones nos indican que estamos ante un dataset desbalanceado, la mayoria de clientes abandonan el servicio.

Esto es un comportamiento normal en servicios de e-commerce a diferencia de servicios tipo SaaS en los que se utilizan servicios de suscripcion y hay sistemas de fidelizacion. En estos servicios el Churn es muy inferior.

## Valores nulos

A parte de las que tenemos, hemos descubierto que tenemos datos nulos en las columnas:
- *'order_approved_at'*: tenemos 14 datos nulos en esta columna, esto puede ser por ordenes canceladas, tendremos que tenerlo en cuenta y comprobarlo, comparando el *'order_status'* con estos valores nulos.
- *'order_delivered_carrier_date'*: 991 datos nulos, esta columna nos indica la fecha de entrega a la empresa de transporte, el motivo de estos datos nulos, puede ser el mismo que el anterior, pedidos cancelados, tendremos que comprobarlo igualmente para saber como imputarlo.
- *'order_delivered_customer_date'*: En este caso tenemos 2087, esta columna representa la fecha de entrega al cliente, este es un dato importante, teniendo en cuenta que el dato aumenta, puede significar, pedidos que no han llegado al cliente, puede ser un indicador de baja satisfaccion que influya en el Churn.

#### Valores nulos *'order_approved_at'*

In [ ]:
df[df['order_approved_at'].isnull()] 

Como son pocas columnas he podido hacer una inspeccion de todas ellas, parece ser que mi idea inicial es incorrecta, los productos han sido entregados aunque no aparezca la fecha de aprobacion. Por lo tanto esta columna no podra aportarmos mayor informacion y la descartare.

#### Valores nulos *'order_delivered_carrier_date'*

In [ ]:
df[df['order_delivered_carrier_date'].isna()].groupby('order_status').size()

He filtrado los datos nulos de esta columna por su order status y los he agrupado, de esta manera puedo intentar obtener informacion de porque existen estos valores nulos.

El resultado es que solo 2 de ellas fueron entregadas. Esto podria ser un indicador de bajo nivel de satisfaccion que podria afectar al Churn

#### Valores nulos *'order_delivered_customer_date'*

In [ ]:
df[df["order_delivered_customer_date"].isna() & df["order_delivered_carrier_date"].isna()] \
    .groupby("order_status").size().sum()

La columnas *'order_delivered_carrier_date'* tiene los datos missing que tiene la columna *'order_delivered_customer_date'*, por lo tanto podre descartar la primera, ya que me da informacion redundante.

Una vez revisado todo esto, mi decision es descartar las dos primeras columnas y me quedarme solo con *'order_delivered_customer_date'*, los datos nulos los tomare como que el producto no ha sido entregado.

La razon de esto es que estas columnas no me aportan informacion extra, la comparacion mas adelante la hare entre la fecha de entrega y la estimada, para saber si hubo algun retraso que pueda afectar a la satisfaccion del cliente

In [ ]:
df.drop(columns=['order_delivered_carrier_date', 'order_approved_at'], inplace=True)

# 

# Analisis Univariante

### Variable Churn

In [ ]:
churn_counts = df["churn"].value_counts()
label_map = {True: "Churn", False: "No churn"}
labels = [label_map[val] for val in churn_counts.index]
palette = ["blue", "orange"]

fig, ax = plt.subplots(1, 2, figsize=(12, 5))

sns.countplot(x="churn", data=df, ax=ax[0], palette=palette, hue='churn', legend=False)
ax[0].set_title("Distribución y conteo por clases de Churn")
ax[0].set_xlabel("Churn")
ax[0].set_ylabel("Número de clientes")
ax[0].set_xticks(range(len(labels_present)))
ax[0].set_xticklabels(reversed(labels))

ax[1].pie(churn_counts, labels=labels, autopct='%1.1f%%', startangle=90, explode=(0.05, 0), shadow={'ox': -0.02}, colors=reversed(palette))
ax[1].set_title("Distribución de Churn")
ax[1].legend(labels, title="Clases", loc="center right", bbox_to_anchor=(0.1, 0.9))

plt.show()

Podemos observar en estos graficos, que mi clase mayoritaria es Churn con un 61% de clientes que abandonan el servicio. Como he comentado antes, en un negocio de e-commerce esta metrica es normal, porque son muchos los clientes que compran algun articulo en especifico y no vuelven a realizar compras.

Tambien podemos observar que son unos 40.000 clientes habituales y unos 60.000 son los que lo han utilizado alguna vez pero no hemos conseguido retener en el servicio.

En nuestro estudio, tener esto en cuenta indispensable para intentar mejorar nuestra estrategia y detectar los patrones de los clientes que usan el servicio habitualmente.

# TODO comentar sobre los 3 limites de fechas de Churn